### PR TITLE
GAWB-741: Handle delayed access to workspace buckets

### PR DESCRIPTION
**Pull request by MatthewBemis**
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Make sure liquibase is updated if appropriate
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] **Submitter**: If PR includes new or changed db queries, include the explain plans in the description
- [x] Tell  that the PR exists if he wants to look at it **(apply requires_doge label)**
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by MattPutnam** [comment]
It's hard for me to comment on Rawls, but everything seems reasonable here.  👍

File: build.sbt [hunk 1]
```diff
@@ -40,7 +40,7 @@ libraryDependencies ++= {
     "org.apache.commons" % "commons-jexl" % "2.1.1",
     ("org.broadinstitute" %% "wdl4s" % "0.4"),
     "org.broadinstitute.dsde.vault" %% "vault-common" % "0.1-15-bf74315",
-    ("com.google.apis" % "google-api-services-storage" % "v1-rev30-1.20.0").exclude("com.google.guava", "guava-jdk5"),
+    ("com.google.apis" % "google-api-services-storage" % "v1-rev35-1.20.0").exclude("com.google.guava", "guava-jdk5"),
```
Comment (MatthewBemis) [inline build.sbt:43]: this did not need to be updated. willing to change it back if anyone cares

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala [hunk 2]
```diff
@@ -462,6 +462,23 @@ class HttpGoogleServicesDAO(
     result
   }
 
+  def diagnosticBucketRead(user: RawlsUser, bucketName: String): Future[Option[ErrorReport]] = {
+    getUserCredentials(user) map { credentialOpt =>
+      credentialOpt match {
+        case None => Some(ErrorReport(StatusCodes.InternalServerError, "Unable to load credentials for user"))
+        case Some(credential) => {
+          val getter = getStorage(credential).objects().list(bucketName)
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala]: note to self: list is most likely not the best thing to do to test if a user can read a bucket (particularly if there are a lot of files in the bucket). if we're calling this endpoint every time a user loads a workspace, we want it to be as quick as possible

File: src/main/resources/swagger/rawls.yaml [hunk 3]
```diff
@@ -2388,6 +2388,44 @@ paths:
             - openid
             - email
             - profile
+  '/api/workspaces/{workspaceNamespace}/{workspaceName}/bucket':
+    get:
+      responses:
+        '204':
```
Comment (MatthewBemis) [inline src/main/resources/swagger/rawls.yaml]: this is actually 200- will fix in next commit

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala [hunk 4]
```diff
@@ -462,6 +462,23 @@ class HttpGoogleServicesDAO(
     result
   }
 
+  def diagnosticBucketRead(user: RawlsUser, bucketName: String): Future[Option[ErrorReport]] = {
+    getUserCredentials(user) map { credentialOpt =>
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala]: this is done on a user's request, don't we already have the user's access token? If we do, I think we should use that.

File: src/main/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiService.scala [hunk 5]
```diff
@@ -78,6 +78,12 @@ trait WorkspaceApiService extends HttpService with PerRequestCreator with UserIn
         }
       }
     } ~
+    path("workspaces" / Segment / Segment / "bucket" ) { (workspaceNamespace, workspaceName) =>
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiService.scala]: I would call this something like testBucketRead

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 6]
```diff
@@ -1205,6 +1207,17 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
     }
   }
 
+  def hasBucketReadAccess(workspaceName: WorkspaceName) = {
+    dataSource.inTransaction { dataAccess =>
+      withWorkspaceContextAndPermissions(workspaceName, WorkspaceAccessLevels.Read, dataAccess) { workspaceContext =>
+        DBIO.from(gcsDAO.diagnosticBucketRead(RawlsUser(RawlsUserSubjectId(userInfo.userSubjectId), RawlsUserEmail(userInfo.userEmail)), workspaceContext.workspace.bucketName)).flatMap {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: use map not flatMap then get rid of the DBIO.successful

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 7]
```diff
@@ -1205,6 +1207,17 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+          case Some(report) => DBIO.successful(RequestComplete(report.statusCode.getOrElse(StatusCodes.InternalServerError)))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: why not just RequestComplete(report)?

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 8]
```diff
@@ -1205,6 +1207,17 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
     }
   }
 
+  def hasBucketReadAccess(workspaceName: WorkspaceName) = {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: change name to testBucketReadAccess

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala [hunk 9]
```diff
@@ -1452,4 +1452,20 @@ class WorkspaceApiServiceSpec extends ApiServiceSpec {
...
+  }
+
+  it should "return 404 when a user can't read the bucket because they dont have workspace access" in withEmptyWorkspaceApiServices("no-access") { services =>
+    Get(s"/workspaces/${testData.workspace.namespace}/${testData.workspace.name}/bucket") ~>
+      sealRoute(services.workspaceRoutes) ~>
+      check {
+        assertResult(StatusCodes.NotFound) { status }
+      }
+  }
+
```
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala:1470]: test also when the google service has a problem... maybe do it to the integration test by manually removing user's access to bucket

**Pull request merged**